### ISSUE TITLE
Adding just the NonMaxSuppression layer

### ISSUE THREAD

**Issue by avasbr** [openvinotoolkit/openvino#1540]
Is there a way to bypass using `DetectionOutput` and use just the `NonMaxSuppression` layer directly when converting detection models? I'm starting from an onnx model where I've already decoded the bounding boxes  and converted the logits to scores. All that's left at this point is NMS, and I would like to not pass in the anchors, variances if I don't need to. 

Are there any examples (or anything close to what I'm looking for) of going from already decoded bboxes and scores and _just_ running NMS with openvino?

**Comment by jgespino** [comment]
Hi @avasbr 

If I understand correctly, you want to change the exit layer of your model (From `DetectionOutput` to `NonMaxSuppression`)? If so, I believe you can use the `--output` parameter in the model optimizer to specify a new exit node. Take a look at the [Cutting off Parts of a Model](https://docs.openvinotoolkit.org/latest/openvino_docs_MO_DG_prepare_model_convert_model_Cutting_Model.html) section in the documentation and let me know if this is what you were looking for. 

Regards,
Jesus

**Comment by avasbr** [comment]
yes, basically i have an onnx model that outputs `scores` and `bboxes`, which are two inputs required by the `NonMaxSuppression` layer. however, there are other parameters such as `iou_threshold` and `score_threshold` which I would also need to pass in, and i don't see how that's possible using just the `--output` arg?

**Comment by jgespino** [comment]
@avasbr I'm not sure this is possible,. However, let me check with some peers and get back to you.

**Comment by avasbr** [comment]
@jgespino thanks for the response. so I ended up converting things properly when I attached a `NonMaxSuppression` layer on the onnx side directly. 

I'd still be curious to find out after you check with your peers whether this is possible to do on the Openvino side, because i'm sure I will run into cases where an op in one framework doesn't translate cleanly (due to mismatches in input/output), and I'll need to know how to do this type of graph surgery. 

Thanks!

**Comment by suygungelen** [comment]
@avasbr  - You can also use add_outputs function inside your application like net.add_outputs('layer_name'). With this function, you can mark any intermediate layer as output layer to retrieve the inference results from the specified layers. For details please check [user guide](https://docs.openvinotoolkit.org/latest/ie_python_api/classie__api_1_1IENetwork.html#ae8024b07f3301d6d5de5c0d153e2e6e6).

**Comment by lazarevevgeny** [comment]
@avasbr, Model Optimizer is not responsible for adding more operations to the graph when converting the model, so you need to update the original framework model and insert needed operations. 

The `add_output` function will not help in this case also, because it allows just to specify the intermediate node of the graph as output and do not allow to add new operations. And it is not possible to decompose the DetectionOutput operation and extract the NMS from it.

However, you can add new operations after loading the network from IR using the nGraph. Refer to this example on how to build the network in runtime: https://docs.openvinotoolkit.org/latest/openvino_docs_IE_DG_nGraphTutorial.html

**Comment by jgespino** [comment]
@avasbr Closing issue, please re-open if additional assistance is needed.

**Comment by ibrahimsoliman97** [comment]
I successfully replaced DetectionOutput Layer with NonMaxSuppression since i has a similar used case as @avasbr , You are required to Edit the subgraph of your model during model optimization. 

Original Code from ObjectDetectionAPI.py at at model_optimizer/extenstions/front/tf/
![image](https://user-images.githubusercontent.com/30038478/93008086-71721e00-f5a3-11ea-8453-32e24a9a8315.png)

Then i edit it to use NMS without Prior, since my network got the classes scores and and location only,  below how can you pass the iou_threshold and other optional parameters..  
![image](https://user-images.githubusercontent.com/30038478/93008107-b4cc8c80-f5a3-11ea-94b0-be61fd1bba93.png)

I hope this is answering your question @avasbr 

Anyway, NonMaxSuppresstion doesn't seem to work as documented [here](https://docs.openvinotoolkit.org/latest/openvino_docs_ops_sort_NonMaxSuppression_3.html)
I am getting correct Output shape, but when i infer in actually image, i got wrong output (not as expected):
![image](https://user-images.githubusercontent.com/30038478/93008187-8dc28a80-f5a4-11ea-8b66-a27479a79c71.png)
So i have 7 classes and 5118 prediction (output of classification head = 1 x 7 x 5118 ) (Output of bbox head = 1 x 5118 x 4)
and i set `max_output_boxes_per_class` for 10, so i got correct output shape of (70x3).

According to NMS documentation, i should get:
ouptu: selected_indices - tensor of type T_IND and shape [min(num_boxes, max_output_boxes_per_class * num_classes), 3] containing information about selected boxes as triplets [batch_index, class_index, box_index]. The output tensor is filled with 0s 
for output tensor elements if the total number of selected boxes is less than the output tensor size.

but when i print my output of NMS, i looks as below (which doesn't make sense for me):
![image](https://user-images.githubusercontent.com/30038478/93008263-56081280-f5a5-11ea-83d7-e69a5a120d5a.png)

Furthermore, if i took my classification and bbox head and threshold them by myself before the NMS Node:
![Detection](https://user-images.githubusercontent.com/30038478/93008309-ca42b600-f5a5-11ea-988f-077f8065cc27.png)
Which include multiple prediction boxes (No NMS), I just trying to validate that my BBox and classification scores are correct.
Now, I hope if anyone can help on validating the NonMaxSuppression op or giving any advice on how to solve this problem..

**Comment by ibrahimsoliman97** [comment]
The error is caused by inferring using MYRIAD Device.
If i used CPU, NMS can work correctly...
Any plan to support NonMaxSuppression Op on MYRIAD ? @jgespino @lazarevevgeny

**Comment by lazarevevgeny** [comment]
@taka-no-me @abotsiev , looks like some issue with the NMS on Myriad. Could you take a look?

**Comment by jgespino** [comment]
Hi @ibrahimsoliman97 this should be fixed in the next software release. Could you share your model and I can verify?

**Comment by avasbr** [comment]
@ibrahimsoliman97 i see, this is definitely closer to what i was looking for, thanks for the explanation. I ended up sticking to modifying the onnx graph directly since that seemed to translate well without issues, but I'll definitely keep this solution in mind in case I need it down the road!
